File: python/src/aoc/year2015/day7.py
from aoc.util import load_example, load_input

M = 2 ** 16


def prepare_rules(lines):
    rules = {}
    for line in lines:
        l, r = line.strip().split(" -> ")
        rules[r] = l.split()
    return rules


def deduct(rules, wire):
    if wire.isnumeric():
        return int(wire)
    lh = rules[wire]
    if not isinstance(lh, list):
        if isinstance(lh, int):
            return lh
        elif lh.isnumeric():
            rules[wire] = int(lh)
        else:
            rules[wire] = deduct(rules, lh)
    elif len(lh) == 1:
        return deduct(rules, lh[0])
    elif len(lh) == 2:
        op, ref = lh
        val = deduct(rules, ref)
        if op == "NOT":
            rules[wire] = M - 1 - val
        else:
            raise Exception
    elif len(lh) == 3:
        ref1, op, ref2 = lh
        val1 = deduct(rules, ref1)
        val2 = deduct(rules, ref2)
        if op == "AND":
            rules[wire] = val1 & val2
        elif op == "OR":
            rules[wire] = val1 | val2
        elif op == "LSHIFT":
            rules[wire] = val1 << val2
        elif op == "RSHIFT":
            rules[wire] = val1 >> val2
        else:
            raise Exception
    else:
        raise Exception
    return rules[wire]


def run(lines, wire, new_rule_b=None):
    """
    >>> data = load_example(__file__, '7')
    >>> run(data, 'd')
    72
    >>> run(data, 'e')
    507
    >>> run(data, 'f')
    492
    >>> run(data, 'g')
    114
    >>> run(data, 'h')
    65412
    >>> run(data, 'i')
    65079
    >>> run(data, 'x')
    123
    >>> run(data, 'y')
    456
    """
    rules = prepare_rules(lines)
    if new_rule_b:
        rules["b"] = new_rule_b
    return deduct(rules, wire)


def part1(lines):
    return run(lines, "a")


def part2(lines):
    answer1 = part1(lines)
    return run(lines, "a", answer1)


if __name__ == "__main__":
    data = load_input(__file__, 2015, "7")
    print(part1(data))
    print(part2(data))

File: rust/src/aoc/year2015/day7.rs
use std::collections::HashMap;
use std::num::ParseIntError;

pub fn yd() -> (usize, usize) {
    (2015, 7)
}

fn foo(wires: &HashMap<&str, u16>, token: &str) -> Result<u16, ParseIntError> {
    match token.parse::<u16>() {
        Ok(value) => Ok(value),
        ParseIntError => {
            if wires.contains_key(token) {
                Ok(wires[token])
            } else {
                ParseIntError
            }
        }
    }
}

fn part1_internal(data: &str, b: u16) -> HashMap<&str, u16> {
    let mut wires: HashMap<&str, u16> = HashMap::new();
    if b > 0 {
        wires.insert("b", b);
    }
    let mut rules: Vec<&str> = data.lines().collect();
    while rules.len() > 0 {
//        println!("{}", rules.len());
        let mut next_rules: Vec<&str> = Vec::new();
        for line in rules.iter() {
            match line {
                // x AND y -> z means that the bitwise AND of wire x and wire y is provided to wire z.
                line if line.contains(" AND ") => {
                    let token: Vec<&str> = line.split(" ").collect();
                    let x = foo(&wires, token.get(0).unwrap());
                    let y = foo(&wires, token.get(2).unwrap());
                    let z = token.get(4).unwrap();
                    if x.is_err() || y.is_err() {
                        next_rules.push(line);
                    } else if !wires.contains_key(z) {
                        wires.insert(z, x.unwrap() & y.unwrap());
                    }
                }
                // x OR y -> z
                line if line.contains(" OR ") => {
                    let token: Vec<&str> = line.split(" ").collect();
                    let x = foo(&wires, token.get(0).unwrap());
                    let y = foo(&wires, token.get(2).unwrap());
                    let z = token.get(4).unwrap();
                    if x.is_err() || y.is_err() {
                        next_rules.push(line);
                    } else  if !wires.contains_key(z) {
                        wires.insert(z, x.unwrap() | y.unwrap());
                    }
                }
                // p LSHIFT 2 -> q means that the value from wire p is left-shifted by 2 and then provided to wire q.
                line if line.contains(" LSHIFT ") => {
                    let token: Vec<&str> = line.split(" ").collect();
                    let p = foo(&wires, token.get(0).unwrap());
                    let value = foo(&wires, token.get(2).unwrap());
                    let q = token.get(4).unwrap();
                    if p.is_err() || value.is_err() {
                        next_rules.push(line);
                    } else  if !wires.contains_key(q) {
                        wires.insert(q, p.unwrap() << value.unwrap());
                    }
                }
                // p RSHIFT 2 -> q
                line if line.contains(" RSHIFT ") => {
                    let token: Vec<&str> = line.split(" ").collect();
                    let p = foo(&wires, token.get(0).unwrap());
                    let value = foo(&wires, token.get(2).unwrap());
                    let q = token.get(4).unwrap();
                    if p.is_err() || value.is_err() {
                        next_rules.push(line);
                    } else  if !wires.contains_key(q) {
                        wires.insert(q, p.unwrap() >> value.unwrap());
                    }
                }
                // NOT e -> f means that the bitwise complement of the value from wire e is provided to wire f.
                line if line.contains("NOT ") => {
                    let token: Vec<&str> = line.split(" ").collect();
                    let e = foo(&wires, token.get(1).unwrap());
                    let f = token.get(3).unwrap();
                    if e.is_err() {
                        next_rules.push(line);
                    } else if !wires.contains_key(f)  {
                        wires.insert(f, !e.unwrap());
                    }
                }
                // 123 -> x means that the signal 123 is provided to wire x.
                _ => {
//                    println!("{line}");
                    let token: Vec<&str> = line.split(" ").collect();
                    let t0 = foo(&wires, token.get(0).unwrap());
                    let x = token.get(2).unwrap();
                    if t0.is_err() {
                        next_rules.push(line);
                    } else if !wires.contains_key(x) {
                        wires.insert(x, t0.unwrap());
                    }
                }
            }
        }
        rules = next_rules;
    }
    wires
}

pub fn part1(data: &str) -> u16 {
    part1_internal(data, 0)["a"]
}

pub fn part2(data: &str) -> u16 {
    let b = part1_internal(data, 0)["a"];
    part1_internal(data, b)["a"]
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    #[test]
    fn tests_part1() {
        const INPUT: &str = "123 -> x
456 -> y
x AND y -> d
x OR y -> e
x LSHIFT 2 -> f
y RSHIFT 2 -> g
NOT x -> h
NOT y -> i";

        let actual = part1_internal(INPUT, 0);
        let expected = HashMap::from([
            ("d", 72),
            ("e", 507),
            ("f", 492),
            ("g", 114),
            ("h", 65412),
            ("i", 65079),
            ("x", 123),
            ("y", 456),
        ]);
        assert_eq!(actual, expected);
    }
}
